### PR TITLE
1.19.1 Update

### PR DESCRIPTION
**Pull request by NoJokeFNA**
### Additions
- [x] Create `ChatCompletionAction` enum
- [x] Create `LastSeenMessages` class
- [x] Create `ClientAckChat` wrapper
- [x] Create `CustomChatCompletions` wrapper
- [x] Create `DeleteChat` wrapper
- [x] Create `PlayerChatHeader` wrapper
- [x] Add new `Serverbound` and `Clientbound` packets

### Changes
- [x] Update the current `ChatType` enums for modern and legacy support
- [x] Change the current `ChatType` handling as 1.19.1 removed some types and added new ones
- [x] The `ClientLoginStart` wrapper now also has a `playerUUID` field which is nullable (due to cross versions support)
- [x] The `ClientChatCommand` and `ClientChatMessage` wrappers now also have `LastSeenMessages` fields which are nullable (due to cross versions support)
- [x] The `ServerChatMessage` wrapper changed so much in 1.19.1 we just decided to recode it. It's now abstracted similarly to the BaseChunk classes, where you cast it to a subclass to access fields available on specific versions. 
- [x] The `ServerServerData` wrapper now also has a `enforceSecureChat` (boolean) field which is always false on legacy versions
- [x] The `ServerSystemChatMessage` wrapper needs an overhaul to support the new `ChatType` enum
  and because 1.19.1 no longer supports overlays we need to take care of that

### Additions
- Added `EQUALS` to ~~`MultiVersion`~~ `VersionComparison` enum
- Added `ClientVersion$is(VersionComparison comparison, ClientVersion targetVersion)`
- Added `PacketWrapper#limitValue(IntFunction<?> function, int limit)` method
- Added `PacketWrapper$readCollection(IntFunction<?> function, Reader<?> reader)` method
- Added `PacketWrapper$writeCollection(Collection<?> collection, Wrtier<?> writer)` method
- Added `PacketWrapper$readSignature` method
- Added `PacketWrapper$writeSignature(SignatureData data)` method
- Added `PacketWrapper$readLastSeenMessagesEntry` method
- Added `PacketWrapper$writeLastMessagesEntry(LastSeenMessages.Entry entry)` method
- Added `PacketWrapper$readLastSeenMessages` method
- Added `PacketWrapper$writeLastSeenMessages(LastSeenMessages.Update update)` method
- Added `PacketWrapper$readList(Reader<?> reader)` method
- Added `PacketWrapper$writeList(List<?> list, Writer<?> writer)` method
- Added missing `DARKNESS` PotionType
- Added missing `EARTH`, `WIND`, `WATER` and `FIRE` PaintingType's
- Added `width` and `height` values to PaintingType enum class

### Changes
- `PacketWrapper$readMap` and `PacketWrapper$writeMap` now uses our FunctionalInterface's (`Reader<?>` & `Writer<?>`) instead of `Function<PacketWrapper<?>, ?>` and `BiFunction<PacketWrapper<?>, ?>`
- Moved ~~`MultiVersion`~~ `VersionComparison` from `api/src/main/java/com/github/retrooper/packetevents/manager/server/ServerVersion.MultiVersion` to `api/src/main/java/com/github/retrooper/packetevents/manager/server/VersionComparison.java`
- Renamed ~~`MultiVersion`~~ `VersionComparison` to `VersionComparison`

### Removals
- Removed `@Experimental` annotation from `ServerVersion$is(VersionComparison comparison, ServerVersion targetVersion)`
- Removed `PacketWrapper$readVarIntList(K key)` method
- Removed `PacketWrapper$writeVarIntList(List<?> list, Consumer<?> writeValue)` method

### CODE REVIEW

**Comment by derklaro** [comment]
If you need some info:
Login (C -> S) Hello (0x00) introduced a new optional uuid (the profile id)
Play (S -> C) SystemChat (0x5F) removed the chat type id and replaced it with a boolean (wether to show it in the overlay or not)

**Comment by NoJokeFNA** [comment]
Already checked your page :P But thanks for the info anyway ^^

**Comment by derklaro** [comment]
Oh, well then 😂

**Comment by NoJokeFNA** [comment]
Are the changes just fine for you? @derklaro

**Comment by NoJokeFNA** [comment]
Alright, thanks!

**Comment by derklaro** [comment]
DinkDonk: 1.19.1-pre3 has a new packet `0x15 - Custom Chat Completions Packet` (S -> C) and the `MSG_COMMAND` chat type was split into `MSG_COMMAND_INCOMING` and `MSG_COMMAND_OUTGOING`

The new packet has an enum "Action" with three possible values `ADD`, `REMOVE` & `SET` and a list of strings to execute the action with. Seems like (by looking at the client code) that these suggestions are used whenever typing something into the chat which isn't a command (they are stored on the client, in-memory). If no custom values are sent it defaults to the online player names. That packet isn't used on the server side yet tho.

**Comment by NoJokeFNA** [comment]
Thanks!

**Comment by derklaro** [comment]
Hello again! Pre4 is here with more changes & 2 new packets:
NEW: 
`0x18 Delete Chat` (S -> C):
   just the signature of the message which was deleted (byte array).

`0x33 Player Chat Header` (S -> C): 
  Seems like the client is now verifing that no messages are lost along the way by veryfing the previous message and the 
  current messages in a chain check. The server sends the header packet for that purpose AFTER sending the actual message to 
  the player.
  
  Structure:
    Nullable: previous message signature (byte array)
    UUID of the message sender
    The signature of the current message (byte array)
    Hash of the chat message (byte array)
    
MODIFIED:
`0x32 Player Chat` (S -> C):
  The chat message:
    Nullable: previous message signature (byte array)
    UUID of the message sender
    The header signature of the current message (byte array)
    The message content (component)
    The message timestamp
    The salt
    Nullable:
      The last seen sender uuid
      The last seen message signature (byte array)
  The Chat Type:
    The chat type registry id (var int)
    The name of the sender (component)
    Nullable: the name of the receiver (currently only set when using the msg or teammsg command)
    
`0x03 Chat Command` (C -> S)
  Also sends the message salt after the message timestamp
  
`0x04 Chat Packet` (C -> S)
  Removed the saltsignature pair, now writes as following:
  The message, string, max length: 256
  The timestamp
  The salt
  The message signature (byte array)
  If the message was a signed preview (boolean)
    
  
(This is a bit much for a friday evening, but hopefully I broke it down so it is helpful for you)

**Comment by NoJokeFNA** [comment]
Holy cow, that's quite some work haha
Thanks for your work though <3

File: api/src/main/java/com/github/retrooper/packetevents/wrapper/play/client/WrapperPlayClientCustomChatCompletions.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+package com.github.retrooper.packetevents.wrapper.play.client;
+
+import com.github.retrooper.packetevents.event.PacketReceiveEvent;
+import com.github.retrooper.packetevents.protocol.chat.ChatCompletionAction;
+import com.github.retrooper.packetevents.protocol.packettype.PacketType;
+import com.github.retrooper.packetevents.wrapper.PacketWrapper;
+
+public class WrapperPlayClientCustomChatCompletions extends PacketWrapper<WrapperPlayClientCustomChatCompletions> {
+    private ChatCompletionAction action;
+    private String[] entries;
```
[the remark this answers is not in the document]
Comment (NoJokeFNA) [reply api/src/main/java/com/github/retrooper/packetevents/wrapper/play/client/WrapperPlayClientCustomChatCompletions.java:28]: Alright!

**Comment by NoJokeFNA** [comment]
IMO if we provide a `readList` method, we should also provide a `writeList` method, but good option in general

**Comment by NoJokeFNA** [comment]
Hold up, we already have such a method...

**Comment by NoJokeFNA** [comment]
Great! haha

**Comment by NoJokeFNA** [comment]
Since I'm not that familiar with the signature, maybe you could give the wrappers a look at what I did and tell me if it's solved right or wrong

**Comment by NoJokeFNA** [comment]
Yea that's exactly what's confusing me lol

**Comment by NoJokeFNA** [comment]
I'm getting an exception while loading the ClientboundPacketType_1_19_1:
https://pastes.dev/QhSTFW2b6Z

File: api/src/main/java/com/github/retrooper/packetevents/protocol/packettype/PacketType.java [hunk 2]
```diff
@@ -521,7 +522,12 @@ public enum Server implements PacketTypeCommon {
             CHAT_PREVIEW_PACKET,
             SERVER_DATA,
             DISPLAY_CHAT_PREVIEW,
-            SYSTEM_CHAT_MESSAGE;
+            SYSTEM_CHAT_MESSAGE,
+
+            //Added in 1.19
```
Comment (NoJokeFNA) [inline api/src/main/java/com/github/retrooper/packetevents/protocol/packettype/PacketType.java:527]:
```suggestion
            //Added in 1.19.1
```

**Comment by derklaro** [comment]
pre5 time:
 - Player Chat Header (S -> C) is now 0x32 (was 0x33)
 - Player Chat (S -> C) is now 0x33 (was 0x32)

New Packet:
 - 0x03 - Chat Ack (C -> S)
   - Collection of Update Entries in the format:
     - UUID of sender
     - Last message header signature (byte array)
   - Optional last received message:
     - UUID of sender
     - Last message header signature (byte array)

Changed packets:    
 - 0x04 (was 0x03) - Chat Command (C -> S) **AND** 0x05 (was 0x04) Chat Packet (C -> S)
    Now contains an update entry (same format as ack) after as the last entry

**Comment by NoJokeFNA** [comment]
Great! - thanks (I hate Mojang)
@retrooper you wanna do it or me?

**Comment by retrooper** [comment]
Please use proper commit names and not just "some progress".

**Comment by NoJokeFNA** [comment]
pre6 didn't change anything packet based

**Comment by derklaro** [comment]
1.19.1-RC2:
  * ServerData (S -> C) now has an additional boolean paramater `enforcesSecureChat` (last byte of the packet)
  * StatusRespnse (S -> C) now includes the same paramater in the json response, the key is called `enforcesSecureChat` as well

**Review by retrooper — APPROVED** [review]
Looking good, ready for 1.19.1!

**Pull request merged**